src/vanilla.ts

```diff
@@ -246,12 +246,12 @@ export type DeepResolveType<T> = T extends (...args: any[]) => any
   : T extends AsRef
   ? T
   : T extends Promise<infer V>
-  ? V
+  ? DeepResolveType<V>
   : T extends object
   ? {
-      [K in keyof T]: DeepResolveType<T[K]>
+      readonly [K in keyof T]: DeepResolveType<T[K]>
     }
-  : T
+  : Readonly<T>
```

Comment:
```suggestion
  : T
```

Looks like we don't need this.
Check this out: https://tsplay.dev/NlxQXN
Or, please let us know the case it might break.

Reply:
Oh yeah that's true, I actually hadn't studied the `Readonly<>` type in detail.
While checking this I noticed that the `Map`, `Set`and `Date` types aren't actually immutable with the current implementation. Do you think we should cover those cases as well, or would that be overkill?

Reply:
They are indeed not immutable. So, it's correct. Actually, it's not deterministic. With #321, we could technically change the behavior at runtime. So, we can't provide 100% static type anyway.